server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -99,95 +206,219 @@ public DeferringBucketCollector getDeferringCollector() {
+        correctRounding(owningBucketOrds);
+        /*
+         * Now that we have the perfect rounding rebucket everything to merge
+         * all of the buckets together that we were too lazy to merge while
+         * collecting.
+         *
+         * TODO it'd be faster if we could apply the merging on the fly as we
+         * replay the hits and build the buckets. How much faster is
+         * *interesting*. Performance tests with a couple of sub-`stats` aggs
+         * show `date_histogram` to have about the same performance as
+         * `auto_date_histogram` so there isn't really much to be gained here.
+         * But if there is a non-delaying but selectivate aggregation "above"
+         * this one then the performance gain could be substantial.
+         */
+        rebucket();
         return buildAggregationsForVariableBuckets(owningBucketOrds, bucketOrds,
                 (bucketValue, docCount, subAggregationResults) ->
                     new InternalAutoDateHistogram.Bucket(bucketValue, docCount, formatter, subAggregationResults),
-                buckets -> {
+                (owningBucketOrd, buckets) -> {
                     // the contract of the histogram aggregation is that shards must return
                     // buckets ordered by key in ascending order
                     CollectionUtil.introSort(buckets, BucketOrder.key(true).comparator());
 
                     // value source will be null for unmapped fields
                     InternalAutoDateHistogram.BucketInfo emptyBucketInfo = new InternalAutoDateHistogram.BucketInfo(roundingInfos,
-                            roundingIdx, buildEmptySubAggregations());
+                            roundingIndexFor(owningBucketOrd), buildEmptySubAggregations());
 
                     return new InternalAutoDateHistogram(name, buckets, targetBuckets, emptyBucketInfo, formatter, metadata(), 1);
                 });
     }
 
+    /**
+     * Pick the correct rounding for the specifies {@code owningBucketOrds}.
+     */
+    private void correctRounding(long[] owningBucketOrds) {
+        for (long owningBucketOrd : owningBucketOrds) {
+            byte oldRounding = roundingIndexFor(owningBucketOrd);
+            if (oldRounding >= roundingInfos.length - 1) {
+                continue;
+            }
+            byte newRounding = (byte)(oldRounding - 1);
+            long count;
+            long min = Long.MAX_VALUE;
+            long max = Long.MIN_VALUE;
+            do {
+                newRounding++;
+                try (LongHash perfect = new LongHash(liveBucketCountUnderestimate.get(owningBucketOrd), context.bigArrays())) {
+                    LongKeyedBucketOrds.BucketOrdsEnum ordsEnum = bucketOrds.ordsEnum(owningBucketOrd);
+                    Rounding.Prepared preparedRounding = preparedRoundings[roundingIndexFor(owningBucketOrd)];
+                    while (ordsEnum.next()) {
```

Comment:
Is it safe to add to the bucketOrds while also iterating over it?  Haven't looked super closely at the internals.

On that note, do we need to clear out the bucketOrdsEnum before adding new keys?  E.g. I'm thinking of the case where we might need to go around this `do...while` loop a few times, and it looks like each time through we'll just keep appending to the ordsEnum so we'll end up with multiple levels of intervals?  I might be misreading though

Reply:
Oh shit! This shld be adding to `perfect`. How did this ever work?!

Reply:
Wow. Ok then. I wonder if I can make a test case that fails because of this......

Reply:
It turns out that the heuristics we had were doing a pretty ok job making the right buckets anyway. So I can just drop this method entirely!